doc/man/polkit-list-actions.1.in
.\" 
.\" polkit-list-actions manual page.
.\" Copyright (C) 2007 David Zeuthen <david@fubar.dk>
.\"
.TH POLKIT-LIST-ACTIONS 1
.SH NAME
polkit-list-actions \- list actions registered with PolicyKit
.SH SYNOPSIS
.PP
.B polkit-list-actions
[options]

.SH DESCRIPTION

\fIpolkit-list-actions\fP simply list the actions that mechanisms have
registered with PolicyKit. For more information about the big picture
refer to the \fIPolicyKit spec\fP which can be found in
.I "@docdir@/spec/polkit-spec.html"
depending on the distribution.

.SH OPTIONS
The following options are supported:
.TP
.I "--help"
Print out usage.
.TP
.I "--version"
Print the version.

.SH BUGS
.PP
Please send bug reports to either the distribution or the HAL
mailing list, see 
.I "http://lists.freedesktop.org/mailman/listinfo/hal"
on how to subscribe.

.SH SEE ALSO
.PP
\&\fIPolicyKit\fR\|(8)

.SH AUTHOR
Written by David Zeuthen <david@fubar.dk> with a lot of help from many
others.



NEWS
==========
PolicyKit 0.4 "Off the Grid"
==========

Released on July 25th 2007.

This is the second release of PolicyKit. 

WARNING WARNING WARNING: do not use this software in a stable
distribution; there are still security sensitive things that needs to
be audited.

Changes from PolicyKit 0.3 "No Man's Land"

David Zeuthen (18):
      post-release version bump
      remove the notion of modules
      replace configuration reload mechanism
      get proper pid and SELinux context
      add support for an /etc/PolicyKit/PolicyKit.conf config file
      document the /etc/PolicyKit/PolicyKit.conf file with a manual page
      update TODO
      switch from VIA_ROOT to VIA_ADMIN
      remove RESULT_NOT_AUTHORIZED_TO_KNOW and s/RESULT_UNKNOWN_ACTION/RESULT_UNKNOWN/
      fix up some of the docs
      change default username
      fix typo
      require that policy files also provide a <message> element
      fix docs
      fix up proper naming of some methods on the PolKitPolicyFileEntry class
      for PolKitContext, mention that a mechanism need to provide .policy files
      update NEWS
      actually reload config and policy files when they change

Requirements for PolicyKit 0.4 ""

 - dbus             >= 0.90
 - glib             >= 2.6.0
 - ConsoleKit       >= 0.2.1
 - expat            >= 1.95.8
 - libselinux       >= 1.30   (optional)

==========
PolicyKit 0.3 "No Man's Land"
==========

Released on June 20th 2007.

This is the first release of PolicyKit. 

WARNING WARNING WARNING: do not use this software in a stable
distribution; there are still security sensitive things that needs to
be audited.

Requirements for PolicyKit 0.3 "No Man's Land"

 - dbus             >= 0.90
 - glib             >= 2.6.0
 - ConsoleKit       >= 0.2.1
 - expat            >= 1.95.8



tools/polkit-list-actions.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 8 -*- */
/***************************************************************************
 *
 * polkit-list-actions.c : list all registered PolicyKit actions
 *
 * Copyright (C) 2007 David Zeuthen, <david@fubar.dk>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307	 USA
 *
 **************************************************************************/

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <stdbool.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <pwd.h>
#include <grp.h>
#include <unistd.h>
#include <errno.h>

#include <polkit/polkit.h>

static void
usage (int argc, char *argv[])
{
	fprintf (stderr,
                 "\n"
                 "usage : polkit-list-actions [--version] [--help]\n"
                 "\n");
	fprintf (stderr,
                 "\n"
                 "        --version        Show version and exit\n"
                 "        --help           Show this information and exit\n"
                 "\n"
                 "List the actions registered with PolicyKit.\n");
}

static void
_print_entry (PolKitPolicyCache *policy_cache,
              PolKitPolicyFileEntry *pfe,
              void *user_data)
{
        const char *action_id;
        const char *group_id;
        PolKitPolicyDefault *def;
        PolKitResult default_inactive;
        PolKitResult default_active;

        action_id = polkit_policy_file_entry_get_id (pfe);
        group_id = polkit_policy_file_entry_get_group_id (pfe);
        def = polkit_policy_file_entry_get_default (pfe);
        default_inactive = polkit_policy_default_get_allow_inactive (def);
        default_active = polkit_policy_default_get_allow_active (def);

        printf ("Policy\n"
                "------\n"
                "group             = %s ('%s')\n"
                "action            = %s ('%s')\n"
                "default_inactive  = %s\n"
                "default_active    = %s\n"
                "\n", 
                group_id, 
                polkit_policy_file_entry_get_group_description (pfe),
                action_id,
                polkit_policy_file_entry_get_action_description (pfe),
                polkit_result_to_string_representation (default_inactive),
                polkit_result_to_string_representation (default_active));
}

int
main (int argc, char *argv[])
{
        int n;

        for (n = 1; n < argc; n++) {
                if (strcmp (argv[n], "--help") == 0) {
                        usage (argc, argv);
                        return 0;
                }
                if (strcmp (argv[n], "--version") == 0) {
                        printf ("polkit-list-actions " PACKAGE_VERSION "\n");
                        return 0;
                }
	}

        int ret;
        ret = 1;

        PolKitContext *ctx;
        PolKitPolicyCache *cache;
        PolKitError *error;
        ctx = polkit_context_new ();
        if (ctx == NULL)
                goto out;
        error = NULL;
        polkit_context_set_load_descriptions (ctx);
        if (!polkit_context_init (ctx, &error)) {
                printf ("Init failed: %s\n", polkit_error_get_error_message (error));
                polkit_context_unref (ctx);
                goto out;
        }

        cache = polkit_context_get_policy_cache (ctx);
        if (cache == NULL) {
                polkit_context_unref (ctx);
                goto out;
        }

        polkit_policy_cache_foreach (cache, _print_entry, NULL);

        polkit_context_unref (ctx);
out:
        return ret;
}
